STYLE.md

```diff
@@ -0,0 +1,262 @@
+large number of dependencies, it can be beneficial to move code into isolated crates under `lib/` in order to allow them
+not only to be shared, but also to reduce the amount of code that must be processed by helper tools like `cargo check`
+and `rust-analyzer` during normal development, which in turn speeds up the feedback loop between writing code and
+getting informed about errors, warnings, and so on.
+
+## `src/`: main binary and all related functionality
+
+The bulk of code resides in the `src/` directory/crate, and within that, is almost entirely related to the
+implementation of the various components -- sources, transforms, and sinks -- that Vector offers. There is also, of
+course, the requisite glue code such as parsing command-line arguments, reading the configuration, constructing and
+configuring components, and wiring them together.
+
+# Internal telemetry: logging, metrics, traces
+
+As a tool for ingesting, transforming, enriching, and shipping observability data, Vector has a significant amount of
+its own internal telemetry. This telemetry is primarily logging and metrics, but also includes some amount of tracing.
+
+## Logging
+
+For logging, we use **[`tracing`](https://docs.rs/tracing/latest/tracing)**, which doubles as both a way to emit
+logs but also a way to use distributed tracing techniques to add nested and contextual metadata by utilizing
+[spans](https://docs.rs/tracing/latest/tracing/#spans).
+
+### Basic Usage
+
+For logging, we use tracing's event macros which should look very similar to almost all other logging libraries, with
+names that emulate the logging level being used i.e. `info!("A wild log has appeared.");`.
+
+All of tracing's event macros -- `trace!`, `debug!`, `info!`, `warn!`, and `error!` -- support the same argument format,
+which allows logging in some common ways:
+
+```rust
+// Basic string literal message, no formatting:
+info!("Server has started.");
+// A formatted message, with the same formatting support as `println!`/`format!`:
+debug!("User connected: {}", username);`
+// Adding structured fields to the even, mixing and matching the message format:
+trace!(bytes_sent = 22, "Sent heartbeat packet to client.")`
+error!(client_addr = %conn.get_ref().peer_addr, "Client actor received malformed packet: {}", parse_err.to_string())
+```
+
+While this does not cover all the permutations of what the macros in `tracing` support, these examples represent the
+preferred style of using the macros.
+
+### Passing in the event message
+
+The `tracing` event macros support passing the event message itself in a few ways, but we prefer the
+**fields/message/message arguments** order:
+
+```rust
+// Don't do this:
+info!(message = "Something happened.");
+// Do this instead:
+info!("Something happened.");
+
+// Don't do this:
+debug!(%client_id, message = "Client entered authentication phase.");
+debug!(message = "Client entered authentication phase.", %client_id);
```

Comment:
Aha, so, first debate about one of my baseline opinions that I shoved in here. :)

My _personal_ feeling is that using `message = "Foo bar doot doot doot..."` is just extra characters to type and clutters up the expression. It looks especially weird when there's no structured data attached, just a "normal" log, as it doesn't match the macros of `log`, even though they can be used interchangeably for non-structured logging.

That said, I didn't realize -- per your `check-events` comment -- that we're looking for instances of `message = "..."` in order to assert correct grammar on them. That does throw a wrench into this particular decree. I'll marinate on this one and see if _maybe_ there's a simple tweak to `check-events` to easily handle my proposed version.

Still interested in thoughts about, ignoring the ability to parse the form in `check-events`, the proposed form and if there's pros/cons, or just opposing personal opinions, that I haven't considered.

Reply:
`check-events` looks for both plain strings and `message = "…"` strings, so that's not the issue per se. The problem is, it only searches at the very front of the log construct for that magic. I _think_ I can adapt the regex to look for either location, but it'll get hairy. It's also worth noting that the actual output puts the message before the structured fields (following the implicit stuff `log` provides, of course).

Reply:
Coming back to this: I haven't yet taken a look at tweaking the regex in `check-events`, but one thing I did look at is using positional format arguments in the message string.

When using the `message = "..."` format, you _cannot_ use positional arguments like you could when using `println!` or `format!`. This lends a little bit of an edge, IMO, to always using the `"message without the field name", ...` approach, because it always works regardless of there are structured fields before it, or positional arguments after it... whereas doing `message = "...", positional_arg` will simply treat `positional_arg` as a structured field, and emit the format string as-is.

Reply:
I've taken a pass at reworking the `check-events` script, specifically the log message validation logic, to accommodate for not only the codebase as it stands now, but also to handle the format I'm proposing in the style guide.